internal/graph/reachableresources.go

```diff
@@ -217,42 +216,70 @@ func (crr *CursoredReachableResources) lookupRelationEntrypoint(
-			return crr.redispatchOrReport(ctx, ci, relationReference, drsm, rg, entrypoint, stream, req, dispatched)
-		})
+	return crr.redispatchOrReportOverDatabaseQuery(
+		ctx,
+		redispatchOverDatabaseConfig{
+			ci:                 ci,
+			reader:             reader,
+			subjectsFilter:     subjectsFilter,
+			sourceResourceType: relationReference,
+			foundResourceType:  relationReference,
+			entrypoint:         entrypoint,
+			rg:                 rg,
+			concurrencyLimit:   crr.concurrencyLimit,
+			parentStream:       stream,
+			parentRequest:      req,
+			dispatched:         dispatched,
+		},
+	)
 }
 
-var queryLimit uint64 = uint64(datastore.FilterMaximumIDCount)
+type redispatchOverDatabaseConfig struct {
+	ci cursorInformation
+
+	reader datastore.Reader
 
-func (crr *CursoredReachableResources) chunkedRedispatch(
+	subjectsFilter     datastore.SubjectsFilter
+	sourceResourceType *core.RelationReference
+	foundResourceType  *core.RelationReference
+
+	entrypoint namespace.ReachabilityEntrypoint
+	rg         *namespace.ReachabilityGraph
+
+	concurrencyLimit uint16
+	parentStream     dispatch.ReachableResourcesStream
+	parentRequest    ValidatedReachableResourcesRequest
+	dispatched       *syncONRSet
+}
+
+func (crr *CursoredReachableResources) redispatchOrReportOverDatabaseQuery(
 	ctx context.Context,
-	ci cursorInformation,
-	reader datastore.Reader,
-	subjectsFilter datastore.SubjectsFilter,
-	resourceType *core.RelationReference,
-	handler func(ctx context.Context, ci cursorInformation, resources dispatchableResourcesSubjectMap) error,
+	config redispatchOverDatabaseConfig,
 ) error {
-	return withDatastoreCursorInCursor(ci, "query-rels",
-		func(queryCursor options.Cursor, ci cursorInformation) (options.Cursor, error) {
-			it, err := reader.ReverseQueryRelationships(
+	return withDatastoreCursorInCursor(ctx, config.ci, "query-rels", config.parentStream, config.concurrencyLimit,
+		// Find the target resources for the subject.
+		func(queryCursor options.Cursor) ([]itemAndPostCursor[dispatchableResourcesSubjectMap], error) {
+			it, err := config.reader.ReverseQueryRelationships(
 				ctx,
-				subjectsFilter,
+				config.subjectsFilter,
 				options.WithResRelation(&options.ResourceRelation{
-					Namespace: resourceType.Namespace,
-					Relation:  resourceType.Relation,
+					Namespace: config.sourceResourceType.Namespace,
+					Relation:  config.sourceResourceType.Relation,
 				}),
 				options.WithSortForReverse(options.BySubject),
 				options.WithAfterForReverse(queryCursor),
-				options.WithLimitForReverse(&queryLimit),
 			)
 			if err != nil {
 				return nil, err
 			}
 			defer it.Close()
 
-			rsm := newResourcesSubjectMap(resourceType)
-			var lastTpl options.Cursor
+			// Chunk based on the FilterMaximumIDCount, to ensure we never send more than that amount of
+			// results to a downstream dispatch.
+			rsm := newResourcesSubjectMap(config.sourceResourceType)
```

Comment:
for a follow-up:  we are potentially returning many subject maps with many tuples. We are running a loop that creates new RSM on demand, and the underlying `map` has a size of zero. This handler is going to cause extra work for GC as the maps grow, although I'll admit I'm not sure what the default size of a map is, unlike slices.

I'd suggest writing a benchmark to observe allocations of current code versus hinting size `datastore.FilterMaximumIDCount` to the RSM factory method.

Reply:
This is actually the code we've used until now in v1.21, slightly reorganized here. I can add a count to the resource subject map though